sockeye/inference.py

```diff
@@ -514,156 +547,164 @@ def translate_nd(self, source: mx.nd.NDArray, source_length: mx.nd.NDArray, buck
-                    # re-normalize hypothesis score by length
-                    hyp_scores[hyp_idx] = (hyp_scores[hyp_idx] + accumulated_scores[hyp_idx] * lengths[hyp_idx]) / (
-                        lengths[hyp_idx] + 1)
-                else:
-                    hyp_scores[hyp_idx][:] = np.inf
-                    hyp_scores[hyp_idx][C.PAD_ID] = accumulated_scores[hyp_idx]
-
-            # get self.beam_size smallest hypotheses
-            # prev_hyp_indices: row indices in hyp_scores.
-            # next_word_ids: column indices in hyp_scores
-            # accumulated_scores: chosen smallest scores in hyp_scores
-            (prev_hyp_indices, next_word_ids), accumulated_scores = sockeye.utils.smallest_k(hyp_scores.asnumpy(),
-                                                                                             self.beam_size, t == 0)
-
-            # select attention according to hypothesis
-            attention_prob_score = attention_prob_score[prev_hyp_indices, :]
-
-            # list of new hypothesis that are now finished
-            new_hyp_finished = [word_id in self.stop_ids for word_id in next_word_ids]
-            new_sequences = [None for _ in range(self.beam_size)]
-            new_attention_lists = [None for _ in range(self.beam_size)]
-            for new_hyp_idx, prev_hyp_idx in enumerate(prev_hyp_indices):
-                if not finished[prev_hyp_idx]:
-                    new_sequences[new_hyp_idx] = sequences[prev_hyp_idx] + [next_word_ids[new_hyp_idx]]
-                    new_attention_lists[new_hyp_idx] = attention_lists[prev_hyp_idx] + [
-                        attention_prob_score[new_hyp_idx, :]]
-                else:
-                    new_sequences[new_hyp_idx] = sequences[prev_hyp_idx]
-                    new_attention_lists[new_hyp_idx] = attention_lists[prev_hyp_idx]
-                lengths[new_hyp_idx] = len(new_sequences[new_hyp_idx])
-
-            finished = new_hyp_finished
-            sequences = new_sequences
-            attention_lists = new_attention_lists
-
-            if all(new_hyp_finished):
+            # (1) obtain next predictions and advance models' state
+            # scores: (beam_size, target_vocab_size)
+            # attention_scores: (beam_size, bucket_key)
+            scores, attention_scores, model_states = self._decode_step(model_states)
+
+            # (2) compute length-normalized accumulated scores in place
+            if t == 0:  # only one hypothesis at t==0
+                scores = scores[:1]
+            else:
+                # renormalize scores by length+1, but not for finished hyps;
+                # their predicted distribution is set to one-hot at C.PAD_ID.
+                scores = (scores + scores_accumulated * lengths) / (lengths + 1)
+                self.pad_dist[:, C.PAD_ID] = scores_accumulated
+                scores = mx.nd.where(finished, self.pad_dist, scores)
+                self.pad_dist += np.inf
```

Comment:
Ok, will rename and I think you are right we can get rid of the self.pad_dist += np.inf since we always overwrite the same complete column in every loop before.

Regarding your suggestion: yeah sure, if we would be in numpy dreamland that would work :) unfortunately this doesn't work for NDArrays (not symbols) afaik.